src/buildercore/threadbare/operations.py

```diff
@@ -0,0 +1,638 @@
+import tempfile
+import contextlib
+import subprocess
+from threading import Timer
+import getpass
+from pssh import exceptions as pssh_exceptions
+import os, sys
+from . import state
+from .common import (
+    PromptedException,
+    merge,
+    subdict,
+    rename,
+    cwd,
+    sudo_wrap_command,
+    pwd_wrap_command,
+    shell_wrap_command,
+)
+from pssh.clients.native import SSHClient as PSSHClient
+import logging
+
+LOG = logging.getLogger(__name__)
+
+
+class SSHClient(PSSHClient):
+    def __deepcopy__(self, memo):
+        # do not copy.deepcopy ourselves or the pssh SSHClient object, just
+        # return a reference to the object (self)
+        # - https://docs.python.org/3/library/copy.html
+        return self
+
+
+class NetworkError(BaseException):
+    """generic 'died while doing something ssh-related' catch-all exception class.
+    calling str() on this exception will return the results of calling str() on the
+    wrapped exception."""
+
+    def __init__(self, wrapped_exception_inst):
+        self.wrapped = wrapped_exception_inst
+
+    def __str__(self):
+        # we have the opportunity here to tweak the error messages to make them
+        # similar to their equivalents in Fabric.
+        # original error messages are still available via `str(excinst.wrapped)`
+        space = " "
+        custom_error_prefixes = {
+            # builder: https://github.com/elifesciences/builder/blob/master/src/buildercore/core.py#L345-L347
+            # pssh: https://github.com/ParallelSSH/parallel-ssh/blob/8b7bb4bcb94d913c3b7da77db592f84486c53b90/pssh/clients/native/parallel.py#L272-L274
+            pssh_exceptions.Timeout: "Timed out trying to connect.",
+            # builder: https://github.com/elifesciences/builder/blob/master/src/buildercore/core.py#L348-L350
+            # fabric: https://github.com/mathiasertl/fabric/blob/master/fabric/network.py#L601-L605
+            # pssh: https://github.com/ParallelSSH/parallel-ssh/blob/2e9668cf4b58b38316b1d515810d7e6c595c76f3/pssh/exceptions.py
+            pssh_exceptions.SSHException: "Low level socket error connecting to host.",
+            pssh_exceptions.SessionError: "Low level socket error connecting to host.",
+            pssh_exceptions.ConnectionErrorException: "Low level socket error connecting to host.",
+        }
```

Comment:
I suspected there was code relying on Fabric's error messages rather than exception types. The exception types were likely easy to find, and these messages match perfectly what we depend on.

Reply:
exactly. A little gross but it gets the job done.